front/components/agent_builder/instructions/extensions/InstructionBlockExtension.tsx

```diff
@@ -37,123 +29,131 @@ declare module "@tiptap/core" {
-        } else if (e.key === "Escape") {
-          e.preventDefault();
-          skipCommitRef.current = true;
-          inputRef.current?.blur();
-        }
-      }}
-      className={cn(
-        "w-fit rounded-lg border-none",
-        "mt-0.5 min-h-5 gap-0.5 rounded-md px-1.5 py-1 text-xs font-medium",
-        "dark:bg-background-night dark:text-foreground-night",
-        "uppercase",
-        "focus-visible:s-border-border-focus dark:focus-visible:s-border-border-focus-night",
-        "focus-visible:s-outline-none focus-visible:s-ring-2",
-        "focus-visible:s-ring-highlight/20 dark:focus-visible:s-ring-highlight/50"
-      )}
-      placeholder={localType}
-    />
-  );
-};
+  // Position inside the middle paragraph (not just between nodes)
+  const targetPos = blockPos + 1 + firstParaSize + 1;
 
-interface InstructionBlockTypeButtonProps {
-  type: string;
-  onClick: () => void;
-}
+  editor.commands.setTextSelection(targetPos);
+  editor.commands.focus();
 
-const InstructionBlockTypeButton: React.FC<InstructionBlockTypeButtonProps> = ({
-  type,
-  onClick,
-}) => {
-  return (
-    <button
-      type="button"
-      onClick={onClick}
-      className="cursor-text"
-      aria-label="Edit instruction type"
-    >
-      <Chip size="mini">{type.toUpperCase()}</Chip>
-    </button>
-  );
-};
+  return true;
+}
 
 const InstructionBlockComponent: React.FC<NodeViewProps> = ({
   node,
+  editor,
+  getPos,
+  selected,
   updateAttributes,
-  deleteNode,
 }) => {
-  const { type } = node.attrs as InstructionBlockAttributes;
-  const [isEditingType, setIsEditingType] = useState(false);
+  const [isCollapsed, setIsCollapsed] = useState(
+    node.attrs.isCollapsed || false
+  );
+
+  // Derive display type directly from content
+  // Empty tags should show as truly empty (no chip)
+  let displayType = "";
+
+  // Only check the first paragraph for the tag
+  if (node.childCount > 0) {
+    const firstChild = node.child(0);
+    if (firstChild.type.name === "paragraph") {
+      const text = firstChild.textContent.trim();
+      // Match opening tag pattern: <tagname> or empty <>
+      const match = text.match(/^<(\w*)>$/);
+      if (match) {
+        displayType = match[1] || ""; // Empty string for <>
+      }
+    }
+  }
+
+  // No fallback - if tag is empty, keep display empty
+  // This gives users clear feedback that the tag is empty
+
+  const handleToggle = (e: React.MouseEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+
+    const newCollapsed = !isCollapsed;
+    setIsCollapsed(newCollapsed);
+
+    // Update the node attribute so ProseMirror knows about the state
+    updateAttributes({ isCollapsed: newCollapsed });
+
+    // Just maintain editor focus if it was already focused
+    // Don't change selection or cursor position
+    if (editor.isFocused) {
+      editor.commands.focus();
+    }
+  };
+
+  const ChevronIcon = isCollapsed ? ChevronRightIcon : ChevronDownIcon;
```

Comment:
OOC: we have a `Collapsible` component in sparkle, not sure what it would take to directly use it?